src/Lecture2.hs

```diff
@@ -1,170 +1,234 @@
-zero, you can stop calculating product and return 0 immediately.
+-- ^ ^^ and before this line. Otherwise the test suite might fail  ^^^
 
->>> lazyProduct [4, 3, 7]
-84
--}
+-- | Implement a function that finds a product of all the numbers in
+-- the list. But implement a lazier version of this function: if you see
+-- zero, you can stop calculating product and return 0 immediately.
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct [] = 1
+lazyProduct (0 : _) = 0
+lazyProduct (h : t) = h * lazyProduct t
+
+-- lazyProduct =
+--   iter id
+--   where
+--     iter k [] = k 1
+--     iter k (h : t) = iter (k . \x -> h * x) t
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate =
+  let dupleIter [] = []
+      dupleIter (h : t) = h : h : dupleIter t
+   in dupleIter
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Integer -> [a] -> (Maybe a, [a])
+removeAt =
+  removeIter (Nothing, [])
+  where
+    removeIter (x, l) _ [] = (x, reverse l)
+    removeIter (x, l) count (h : t)
+      | count < 0 = (Nothing, h : t)
+      | count == 0 = (Just h, reverse l ++ t)
+      | otherwise = removeIter (x, h : l) (count - 1) t
+
+-- | Write a function that takes a list of lists and returns only
+-- lists of even lengths.
+--
+-- >>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
+-- [[3,1,2,7],[]]
+--
+-- ♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
+--  in this function.
+evenLists :: [[a]] -> [[a]]
+evenLists =
+  let evenfilter l = [i | i <- l, (even . length) i]
+   in evenfilter
+
+-- | The @dropSpaces@ function takes a string containing a single word
+-- or number surrounded by spaces and removes all leading and trailing
+-- spaces.
+--
+-- >>> dropSpaces "   hello  "
+-- "hello"
+-- >>> dropSpaces "-200            "
+-- "-200"
+--
+-- ♫ NOTE: As in the previous task, use eta-reduction and function
+--  composition (the dot (.) operator) in this function.
+--
+-- 🕯 HINT: look into Data.Char and Prelude modules for functions you may use.
+dropSpaces :: [Char] -> [Char]
+dropSpaces =
+  spacefilter False []
+  where
+    spacefilter _ acc [] = reverse acc
+    spacefilter False acc (h : t)
+      | isSpace h = spacefilter False acc t
+      | otherwise = spacefilter True (h : acc) t
+    spacefilter True acc (h : t)
+      | isSpace h = reverse acc
+      | otherwise = spacefilter True (h : acc) t
```

Comment:
That's the correct solution 👍🏻 
You can also use the standard function [dropWhile](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:dropWhile) and `takeWhile` with the combination of `isSpace` to write `dropSpaces` as a one-liner 👍🏻